packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,263 @@
+import * as React from 'react';
+import { unstable_batchedUpdates } from 'react-dom';
+import { HTMLElementWithStyledMap, getMotionDuration } from '../utils/dom-style';
+import { useAnimationFrame, useTimeout, usePrevious } from '@fluentui/react-utilities';
+
+export type UseMotionOptions = {
+  /**
+   * Whether to animate the element on first mount.
+   *
+   * @default false
+   */
+  animateOnFirstMount?: boolean;
+};
+
+export type MotionType = 'unmounted' | 'entering' | 'entered' | 'idle' | 'exiting' | 'exited';
+
+export type MotionState<Element extends HTMLElement = HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.Ref<Element>;
+
+  /**
+   * Current state of the element.
+   *
+   * - `unmounted` - The element is not yet rendered or can be safely removed from the DOM.
+   * - `entering` - The element is performing enter animation.
+   * - `entered` - The element has finished enter animation.
+   * - `idle` - The element is currently not animating, but rendered on screen.
+   * - `exiting` - The element is performing exit animation.
+   * - `exited` - The element has finished exit animation.
+   */
+  type: MotionType;
+
+  /**
+   * Indicates whether the component is currently rendered and visible.
+   * Useful to apply CSS transitions only when the element is active.
+   */
+  isActive(): boolean;
+
+  /**
+   * Indicates whether the component can be rendered.
+   * This can be used to avoid rendering the component when it is not visible anymore.
+   */
+  canRender(): boolean;
+};
+
+export type MotionShorthandValue = boolean;
+
+export type MotionShorthand<Element extends HTMLElement = HTMLElement> = MotionShorthandValue | MotionState<Element>;
+
+/**
+ * @internal
+ *
+ * Hook to manage the presence of an element in the DOM based on its CSS transition/animation state.
+ *
+ * @param present - Whether the element should be present in the DOM
+ * @param events - Callbacks for when the element enters or exits the DOM
+ */
+function useMotionPresence<Element extends HTMLElement>(
+  presence: boolean,
+  options: UseMotionOptions = {},
+): MotionState<Element> {
+  const { animateOnFirstMount } = { animateOnFirstMount: false, ...options };
+
+  const [type, setType] = React.useState<MotionType>(presence ? 'idle' : 'unmounted');
+  const [active, setActive] = React.useState<boolean>(false);
+
+  const [currentElement, setCurrentElement] = React.useState<HTMLElementWithStyledMap<Element> | null>(null);
+  const [setAnimationTimeout, clearAnimationTimeout] = useTimeout();
+  const [setActiveAnimationFrame, cancelActiveAnimationFrame] = useAnimationFrame();
+  const [setProcessingAnimationFrame, cancelProcessingAnimationFrame] = useAnimationFrame();
+  const [setDelayedAnimationFrame, cancelDelayedAnimationFrame] = useAnimationFrame();
+  const skipAnimationOnFirstRender = React.useRef(!animateOnFirstMount);
```

Comment:
nit: perhaps you can achieve the same by using `useFirstMount` hook + `animateOnFirstMount`?

Reply:
That is a good idea. I'll check the implementation.